lib/store/inventory.ex

```diff
@@ -309,11 +309,69 @@ defmodule Store.Inventory do
       if order_product != nil do
         {:error, "Product already in cart"}
       else
-        create_order_product(%{order_id: order.id, product_id: product.id})
+        {val, _} = Integer.parse(product_params["quantity"])
+        size = product_params["size"]
+
+        if update_stock_sizes(product, size, val) == {:error, "Not enough stock"} do
+          {:error, "Not enough stock"}
+        else
+          create_order_product(%{
+            order_id: order.id,
+            product_id: product.id,
+            quantity: val,
+            size: size
+          })
+        end
       end
     else
       {:ok, order} = create_order(%{user_id: user.id})
-      create_order_product(%{order_id: order.id, product_id: product.id})
+      {val, _} = Integer.parse(product_params["quantity"])
+      size = product_params["size"]
+
+      if update_stock_sizes(product, size, val) == {:error, "Not enough stock"} do
+        {:error, "Not enough stock"}
+      else
+        create_order_product(%{
+          order_id: order.id,
+          product_id: product.id,
+          quantity: val,
+          size: size
+        })
+      end
+    end
+  end
+
+  def update_stock_sizes(product, size, quantity) do
+    new_sizes =
+      case size do
+        "XS" -> Map.put(product.sizes, :xs_size, product.sizes.xs_size - quantity)
+        "S" -> Map.put(product.sizes, :s_size, product.sizes.s_size - quantity)
+        "M" -> Map.put(product.sizes, :m_size, product.sizes.m_size - quantity)
+        "L" -> Map.put(product.sizes, :l, product.sizes.l_size - quantity)
+        "XL" -> Map.put(product.sizes, :xl_size, product.sizes.xl_size - quantity)
+        "XXL" -> Map.put(product.sizes, :xxl_size, product.sizes.xxl_size - quantity)
+      end
+
+    values = Map.values(new_sizes) |> Enum.filter(&is_integer/1)
+
+    if Enum.any?(values, &(&1 < 0)) or product.stock - quantity < 0 do
```

Comment:
If there's no stock shouldn't we compare to 0 (`== 0`)?

Reply:
I did the calculus previously between the lines 347 and 352 so if any of those values is less than 0 than i put an error

Reply:
Right, but can you show me a case where a value could be less than 0? If everything goes well, "no stock" should be the same as `stock == 0`.

Reply:
Let's say that there were initially 2 shirts in stock, and I purchased both of them, resulting in a quantity of 0 and leaving no more stock available for anyone else, the ```stock will be == 0``` and i can get both pencils. 
Now that's imagine that ```stock == 0``` and i bought 1 shirt , ```values``` array will be like [10,2,3,5,-1,2,3] because the subtraction was previously made between lines ```347 and 352``` and will raise an error.